costs/es_request_parser.go

```diff
@@ -0,0 +1,128 @@
+
+package costs
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"gopkg.in/olivere/elastic.v5"
+)
+
+// flattenSubAggregation allows to transform an array
+// of map to one flattened map.
+func flattenSubAggregation(in *interface{}) {
+	oldTab := (*in).([]interface{})
+	newMap := make(map[string]interface{})
+	for _, field := range oldTab {
+		for name, subField := range field.(map[string]interface{}) {
+			newMap[name] = subField
+		}
+	}
+	*in = newMap
+}
+
+// stripAggregationArchitecture strips an array by
+// deleting some values and calling flattenSubAggregation.
+func stripAggregationArchitecture(aggrName string, aggrType interface{}, in map[string]interface{}) {
+	key, ok := in["key_as_string"]
+	if !ok {
+		key, ok = in["key"]
+	}
+	if skey := fmt.Sprintf("%v", key); ok {
+		if aggrMapType, ok := aggrType.(map[string]interface{}); ok && aggrName[0] == '*' {
+			aggrType := aggrMapType["buckets"]
+			flattenSubAggregation(&aggrType)
+		}
+		in[skey] = aggrType
+		delete(in, "key")
+		delete(in, "key_as_string")
+	}
+	delete(in, aggrName)
+}
+
+// browseMapRecursivly browses a map and calls browseMapRecursivly or
+// browseTabRecursivly, and stripAggregationArchitecture for each fields.
+func browseMapRecursivly(in map[string]interface{}) {
+	for aggrName, field := range in {
+		if aggrName[0] != '&' && aggrName[0] != '*' {
+			continue
+		}
+		field := field.(map[string]interface{})
+		metadata := strings.Split(aggrName[1:], ".")
+		aggrTypeName := strings.Replace(metadata[0], ",", ".", -1)
+		aggrType := field[aggrTypeName]
+		if aggrName[0] == '&' {
+			browseTabRecursivly(&aggrType)
+		} else if aggrName[0] == '*' {
+			browseMapRecursivly(field)
+		}
+		stripAggregationArchitecture(aggrName, aggrType, in)
+	}
+}
+
+// browseTabRecursivly browses a map and calls browseMapRecursivly.
+// In the end it calls flattenSubAggregation to flat the array.
+func browseTabRecursivly(in *interface{}) {
+	tab := (*in).([]interface{})
+	for _, field := range tab {
+		field := field.(map[string]interface{})
+		delete(field, "doc_count")
+		browseMapRecursivly(field)
+	}
+	flattenSubAggregation(in)
+}
+
+// prepareRecursiveParsing prepares the first run of the recursive and
+// calls browseMapRecursivly or browseTabRecursivly.
+func prepareRecursiveParsing(result map[string]interface{}, aggrName string, aggr *json.RawMessage) error {
+	var root map[string]interface{}
+	err := json.Unmarshal(*aggr, &root)
+	if err != nil {
+		return err
+	}
+	metadata := strings.Split(aggrName[1:], ".")
+	name := metadata[len(metadata)-1]
+	aggrTypeName := strings.Replace(metadata[0], ",", ".", -1)
+	aggrType := root[aggrTypeName]
+	if aggrName[0] == '&' {
+		browseTabRecursivly(&aggrType)
+		result[name] = aggrType
+	} else if aggrName[0] == '*' {
+		browseMapRecursivly(root)
+		bucketAggr := aggrType.(map[string]interface{})["buckets"]
+		flattenSubAggregation(&bucketAggr)
+		result[name] = bucketAggr
+	} else {
+		result[aggrName] = root
+	}
```

Comment:
This code here could be summed up to : 
```
/*
Magic happens
*/
```
Either use more explicit variable naming, comment inside the function, or redo the way the logic works.